File: cp-essentials/Linear-Recurrences-And-Matrix-Exponentiation/binary_exponentiation.cpp
#include <bits/stdc++.h>
using namespace std;


int powr(int a, int b) {
	int res = 1;
	while (b) {
		if (b & 1)
			res *= a;
		a *= a;
		b >>= 1;
	}
	return res;
}

int main(int argc, char const *argv[]) {
	cout << powr(2, 30);
	return 0;
}

File: cp-essentials/Linear-Recurrences-And-Matrix-Exponentiation/binary_exponentiation.java

class Main {

	int powr(int a, int b) {
		int res = 1;
		while(b > 0) {
			if(b  1) res *= a;
			a *= a;
			b <<= 1;
		}
		return res;
	}

	public static void main(String[] args) {
		System.out.println(2, 10);
	}
}